esx_service/vmdk_ops.py

```diff
@@ -1703,6 +1704,101 @@ def wait_for_tasks(si, tasks):
 
 #------------------------
 
+def start_vm_changelistener():
+    """
+    Listen to power state changes of VMs running on current host
+    """
+    si = get_si()
+    pc = si.content.propertyCollector
+    create_vm_powerstate_filter(pc, si.content.rootFolder)
+    # Start a separate thread to listen to changes
+    threadutils.start_new_daemon(target=listen_vm_propertychange,
+                args=(pc, []))
+
+def create_vm_powerstate_filter(pc, from_node):
+    """
+    Create a filter spec to list to VM power state changes
+    """
+    filterSpec = vmodl.query.PropertyCollector.FilterSpec()
+    objSpec = vmodl.query.PropertyCollector.ObjectSpec(obj=from_node,
+                                                       selectSet=vm_folder_traversal())
+    filterSpec.objectSet.append(objSpec)
+    # Add the property specs
+    propSpec = vmodl.query.PropertyCollector.PropertySpec(type=vim.VirtualMachine, all=False)
+    propSpec.pathSet.append('runtime.powerState')
+    filterSpec.propSet.append(propSpec)
+    try:
+        pcFilter = pc.CreateFilter(filterSpec, True)
+        atexit.register(pcFilter.Destroy)
+    except Exception as e:
+        logging.error("Problem creating PropertyCollector %s", str(e))
+
+def listen_vm_propertychange(pc, empty):
+    logging.info("PropertyChangeListener thread started")
+    version = ''
+    while True:
+        result = pc.WaitForUpdates(version)
+
+        # if the wait timed out, wait again
+        if result is None:
+            continue
+
+        try:
+            # process the updates result
+            for filterSet in result.filterSet:
+                for objectSet in filterSet.objectSet:
+                    moref = getattr(objectSet, 'obj', None)
+                    if objectSet.kind == 'modify':
+                        for change in objectSet.changeSet:
+                            if change.name == 'runtime.powerState' and change.val == 'poweredOff':
+                                vm_name = moref.config.name
+                                logging.info("VM poweroff change found for %s", vm_name)
+                                # if the event was powerOff for a VM, set the status of all
+                                # docker volumes attached to the VM to be detached
+                                set_device_detached(moref.config.hardware.device)
+        except Exception as e:
+            logging.error("PropertyChangeListener: error %s", str(e))
```

Comment:
Since this is a background operation, we can log an error. Not sure how else we can notify the admin.